### PR TITLE
Arm64: Pass promoted struct using GT_FIELD_LIST

### PR DESCRIPTION
**Pull request by CarolEidt**

### CODE REVIEW

**Comment by CarolEidt** [comment]
Fixes #16377

**Comment by CarolEidt** [comment]
@sdmaclea @dotnet/jitcontrib PTAL
This was extracted from my work to unify and improve struct passing.

**Comment by CarolEidt** [comment]
ping @dotnet/jit-contrib (I misspelled jit-contrib in the previous comment)

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -4962,21 +4978,29 @@ GenTree* Compiler::fgMorphMultiregStructArg(GenTree* arg, fgArgTabEntry* fgEntry
...
-            arg = gtNewOperNode(GT_ADDR, TYP_I_IMPL, arg);
-
-            // Create an Obj of the temp to use it as a call argument.
-            arg = gtNewObjNode(lvaGetStruct(lcl->gtLclNum), arg);
         }
         if (lcl != nullptr)
         {
-            // Its fields will need to accessed by address.
-            lvaSetVarDoNotEnregister(lcl->gtLclNum DEBUG_ARG(DNER_IsStructArg));
+            unsigned lclNum = lcl->gtLclNum;
```
Comment (mikedn) [inline src/jit/morph.cpp]: This local variable doesn't seem necessary

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -5474,6 +5484,43 @@ GenTree* Compiler::fgMorphMultiregStructArg(GenTree* arg, fgArgTabEntry* fgEntry
...
+//
+// Return value:
+//    The new GT_FIELD_LIST that we have created.
+//
+GenTreeFieldList* Compiler::fgMorphLclArgToFieldlist(GenTreeLclVarCommon* lcl)
+{
+    LclVarDsc* varDsc = &(lvaTable[lcl->gtLclNum]);
+    assert(varDsc->lvPromoted == true);
+
+    unsigned          lclNum      = lcl->gtLclNum;
```
Comment (mikedn) [inline src/jit/morph.cpp]: Unused?

File: src/jit/morph.cpp [hunk 3]
```diff
@@ -4962,21 +4978,29 @@ GenTree* Compiler::fgMorphMultiregStructArg(GenTree* arg, fgArgTabEntry* fgEntry
...
+            }
+            else
+            {
+                if (!arg->OperIs(GT_OBJ))
+                {
+                    // Create an Obj of the temp to use it as a call argument.
+                    arg = gtNewOperNode(GT_ADDR, TYP_I_IMPL, arg);
+                    arg = gtNewObjNode(lvaGetStruct(lcl->gtLclNum), arg);
+                }
+                // Its fields will need to accessed by address.
```
Comment (mikedn) [inline src/jit/morph.cpp]: to **be** accessed?

**Comment by sdmaclea** [comment]
test Ubuntu arm64 Cross Checked Innerloop Build and Test

**Comment by sdmaclea** [comment]
@CarolEidt  Thanks

**Comment by CarolEidt** [comment]
@sdmaclea - I made some updates to address @mikedn 's feedback and tried to re-trigger "test Ubuntu arm64 Cross Checked Innerloop Build and Test", but it doesn't seem to have triggered - perhaps I'm not on whatever list that requires.

**Comment by sdmaclea** [comment]
@CarolEidt It didn't trigger for me either.  It might not exist yet or I spelled it wrong...

**Comment by sdmaclea** [comment]
@dotnet-bot help

**Comment by CarolEidt** [comment]
Hmmm - seems to have magically triggered now. Perhaps one just needs to ask for help ;-)

**Comment by jashook** [comment]
@CarolEidt can you add the regression test to the arm64 lst file? I opened a pr against this pr which I believe should be sufficient see https://github.com/CarolEidt/coreclr/pull/1 for more info.

File: tests/src/JIT/Regression/JitBlue/GitHub_16377/GitHub_16377.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+using System;
```
Comment (jashook) [inline tests/src/JIT/Regression/JitBlue/GitHub_16377/GitHub_16377.cs]: I think you need the standard header above.

**Comment by CarolEidt** [comment]
test Ubuntu arm64 Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
test Windows_NT x64 Formatting

**Comment by jashook** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test

File: tests/arm64/Tests.lst [hunk 5]
```diff
@@ -90787,3 +90787,11 @@ Expected=0
 MaxAllowedDurationSeconds=600
 Categories=EXPECTED_PASS
 HostStyle=0
+
+[GitHub_16377.cmd_11720]
+RelativePath=JIT\Regression\JitBlue\GitHub_16377\GitHub_16377.cmd
```
Comment (jashook) [inline tests/arm64/Tests.lst]: Looks like this should be JIT\Regression\JitBlue\GitHub_16377\GitHub_16377\GitHub_16377.cmd sorry

Comment (jashook) [reply tests/arm64/Tests.lst]: https://github.com/CarolEidt/coreclr/pull/2 to address it.

**Comment by CarolEidt** [comment]
@dotnet/dnceng - is there a known problem with the CI system? Jobs aren't triggering, and I've had a couple of process-seeming failures.

**Comment by CarolEidt** [comment]
@mattgal - after the latest push, only 7 jobs seem to have triggered. The arm64 jobs, for example, didn't pass in the last push, and need to re-run. But normally they would get re-triggered with a new commit.

**Comment by MattGal** [comment]
Got it, created https://github.com/dotnet/core-eng/issues/3447 to investigate.

**Comment by CarolEidt** [comment]
test this please

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by CarolEidt** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu arm64 Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu arm64 Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
@jashook can you help me figure out how to trigger the "@Ubuntu arm64 Cross Checked Innerloop Build and Test" job?

**Comment by CarolEidt** [comment]
I'm not sure that stress testing would check anything really useful for struct passing. @sdmaclea is there anything additional you think I should run?

**Comment by sdmaclea** [comment]
You could run corefx tests, but I do not think those are enabled for arm64 yet. #17764 is trying to enable them.

If you want me to pull your PR and run corefx, let me know, but it looked like your test cases should cover everything.

**Comment by CarolEidt** [comment]
FYI - this results in the following diffs, using protononjit:
```
Total bytes of diff: -16984 (-0.01% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
        -400 : JIT\HardwareIntrinsics\X86\Sse2\SetVector128_ro\SetVector128_ro.dasm (-0.09% of base)
        -400 : JIT\HardwareIntrinsics\X86\Sse2\SetZeroVector128_ro\SetZeroVector128_ro.dasm (-0.09% of base)
        -336 : JIT\HardwareIntrinsics\X86\Sse2\Shuffle_ro\Shuffle_ro.dasm (-0.07% of base)
        -320 : JIT\HardwareIntrinsics\X86\Sse2\AddSaturate_ro\AddSaturate_ro.dasm (-0.08% of base)
        -320 : JIT\HardwareIntrinsics\X86\Sse2\ShuffleHigh_ro\ShuffleHigh_ro.dasm (-0.07% of base)

63 total files with size differences (63 improved, 0 regressed), 2508 unchanged.

Top method improvements by size (bytes):
         -88 : JIT\Methodical\structs\StructStackParams\StructStackParams.dasm - MyProgram:TestStruct4():int (2 methods)
         -80 : JIT\opt\FastTailCall\FastTailCallCandidates\FastTailCallCandidates.dasm - FastTailCallCandidates:CallerHFaCaseCalleeStackArgs(double,double,double,double,double,double,double,double,double,double):int (2 methods)
         -72 : JIT\opt\FastTailCall\FastTailCallCandidates\FastTailCallCandidates.dasm - FastTailCallCandidates:Tester(int):int (2 methods)
         -64 : JIT\HardwareIntrinsics\X86\Sse2\AddSaturate_ro\AddSaturate_ro.dasm - IntelHardwareIntrinsicTest.TestTableTuvImmSse2`3[__Canon,__Canon,__Canon][System.__Canon,System.__Canon,System.__Canon]:GetDoubleImmDataPoint(int):struct:this (4 methods)
         -64 : JIT\HardwareIntrinsics\X86\Sse2\AddScalar_ro\AddScalar_ro.dasm - IntelHardwareIntrinsicTest.TestTableTuvImmSse2`3[__Canon,__Canon,__Canon][System.__Canon,System.__Canon,System.__Canon]:GetDoubleImmDataPoint(int):struct:this (4 methods)

509 total methods with size differences (509 improved, 0 regressed), 326060 unchanged.
```
The majority of the methods are in the X86 HardwareIntrinsics tests, where, instead of `ldr;ldrsw;str;str` we have `ldp;stp` - this is because, since the argument is address-taken anyway, we use a `GT_OBJ`. However, there are a few cases, such as `Microsoft.CodeAnalysis.CSharp.Symbols.ConstraintsHelper:CheckTypeConstraints` where the code is improved because we no longer mark a stack-passed struct as `lvDoNotEnregister`.

**Pull request merged**